src/lib/support/PoolWrapper.h

```diff
@@ -0,0 +1,112 @@
+    virtual void ReleaseObject(U * element)                                = 0;
+    virtual void ResetObject(U * element, ConstructorArguments &&... args) = 0;
+
+    template <typename Function>
+    bool ForEachActiveObject(Function && function)
+    {
+        auto proxy = [&](U * target) -> bool { return function(target); };
+        return ForEachActiveObjectInner(
+            &proxy, [](void * context, U * target) -> bool { return (*static_cast<decltype(proxy) *>(context))(target); });
+    }
+
+protected:
+    using Lambda                                                         = bool (*)(void *, U *);
+    virtual bool ForEachActiveObjectInner(void * context, Lambda lambda) = 0;
+};
+
+template <class T, size_t N, typename Interface>
+class PoolProxy;
+
+template <class T, size_t N, typename U, typename... ConstructorArguments>
+class PoolProxy<T, N, std::tuple<U, ConstructorArguments...>> : public PoolInterface<U, ConstructorArguments...>
+{
+public:
+    static_assert(std::is_base_of<U, T>::value, "Interface type is not derived from Pool type");
+
+    PoolProxy() {}
+    virtual ~PoolProxy() override {}
+
+    virtual U * CreateObject(ConstructorArguments &&... args) override
+    {
+        return Impl().CreateObject(std::forward<ConstructorArguments>(args)...);
+    }
+
+    virtual void ReleaseObject(U * element) override { Impl().ReleaseObject(static_cast<T *>(element)); }
+
+    virtual void ResetObject(U * element, ConstructorArguments &&... args) override
+    {
+        return Impl().ResetObject(static_cast<T *>(element), std::forward<ConstructorArguments>(args)...);
+    }
+
+protected:
+    virtual bool ForEachActiveObjectInner(void * context,
+                                          typename PoolInterface<U, ConstructorArguments...>::Lambda lambda) override
+    {
+        return Impl().ForEachActiveObject([&](T * target) { return lambda(context, static_cast<U *>(target)); });
+    }
+
+    virtual BitMapObjectPool<T, N> & Impl() = 0;
+};
+
+/*
+ * @brief
+ *   Define a implementation of a pool which derive and expose PoolInterface's.
+ *
+ *  @tparam T          a subclass of element to be allocated.
+ *  @tparam N          a positive integer max number of elements the pool provides.
+ *  @tparam Interfaces a list of parameters which defines PoolInterface's. each interface is defined by a
+ *                     std::tuple<U, ConstructorArguments...>. The PoolImpl is derived from every
+ *                     PoolInterface<U, ConstructorArguments...>, the PoolImpl can be converted to the interface type
+ *                     and passed around
+ */
+template <class T, size_t N, typename... Interfaces>
+class PoolImpl : public PoolProxy<T, N, Interfaces>...
+{
+public:
+    PoolImpl() {}
+    virtual ~PoolImpl() override {}
+
+protected:
+    virtual BitMapObjectPool<T, N> & Impl() override { return mImpl; }
+
+private:
+    BitMapObjectPool<T, N> mImpl;
```

Comment:
In a followup, could you update this to support any Pool type, i.e. also `HeapObjectPool`? `PoolInterface` would be useful for the EndPoint pools in `InetLayer`.

Reply:
No problem, I'll do it in a separate PR.